operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+}
+
+// SetClusterStatus sets the status of the given cluster.
+func (r *ResourceClient[T, U]) SetClusterStatus(cluster U, status ClusterStatus) bool {
+	return r.statusUpdater.Update(cluster, status)
+}
+
+type gvkObject struct {
+	gvk schema.GroupVersionKind
+	nn  types.NamespacedName
+}
+
+// SyncAll synchronizes the simple resources associated with the given cluster,
+// cleaning up any resources that should no longer exist.
+func (r *ResourceClient[T, U]) SyncAll(ctx context.Context, owner U) error {
+	// we don't sync node pools here
+	resources, err := r.listAllOwnedResources(ctx, owner, false)
+	if err != nil {
+		return err
+	}
+	toDelete := map[gvkObject]client.Object{}
+	for _, resource := range resources {
+		toDelete[gvkObject{
+			gvk: resource.GetObjectKind().GroupVersionKind(),
+			nn:  client.ObjectKeyFromObject(resource),
+		}] = resource
+	}
+
+	toSync, err := r.simpleResourceRenderer.Render(ctx, owner)
+	if err != nil {
+		return err
+	}
+
+	// attempt to create as many resources in one pass as we can
+	errs := []error{}
+
+	for _, resource := range toSync {
+		if err := r.patchOwnedResource(ctx, owner, resource); err != nil {
+			errs = append(errs, err)
+		}
+		delete(toDelete, gvkObject{
+			gvk: resource.GetObjectKind().GroupVersionKind(),
+			nn:  client.ObjectKeyFromObject(resource),
+		})
+	}
+
+	for _, resource := range toDelete {
+		if err := r.client.Delete(ctx, resource); err != nil {
+			if !k8sapierrors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// FetchExistingAndDesiredPools fetches the existing and desired node pools for a given cluster, returning
+// a tracker that can be used for determining necessary operations on the pools.
+func (r *ResourceClient[T, U]) FetchExistingAndDesiredPools(ctx context.Context, cluster U) (*PoolTracker, error) {
+	pools := NewPoolTracker(cluster.GetGeneration())
+
+	existingPools, err := r.fetchExistingPools(ctx, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("fetching existing pools: %w", err)
+	}
+
+	desired, err := r.nodePoolRenderer.Render(ctx, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("constructing desired pools: %w", err)
+	}
+
+	pools.addExisting(existingPools...)
+	pools.addDesired(desired...)
+
+	return pools, nil
+}
+
+// Builder is an interface for our used methods of *sigs.k8s.io/controller-runtime/pkg/builder.Builder
+type Builder interface {
+	For(object client.Object, opts ...builder.ForOption) *builder.Builder
+	Owns(object client.Object, opts ...builder.OwnsOption) *builder.Builder
+	Watches(object client.Object, eventHandler handler.EventHandler, opts ...builder.WatchesOption) *builder.Builder
+}
+
+// WatchResources configures resource watching for the given cluster, including StatefulSets and other resources.
+func (r *ResourceClient[T, U]) WatchResources(builder Builder, cluster U) error {
+	// set that this is for the cluster
+	builder.For(cluster)
```

Comment:
Does builder mutate the instance you call the methods on? Seems like this should be `builder = builder.For(cluster)`